src/mem/ruby/protocol/MESI_Three_Level-msg.sm

```diff
@@ -81,8 +81,11 @@ structure(CoherenceMsg, desc="...", interface="Message") {
   PrefetchBit Prefetch,         desc="Is this a prefetch request";
 
   bool functionalRead(Packet *pkt) {
-    // Only PUTX messages contains the data block
-    if (Class == CoherenceClass:PUTX) {
+    // Valid data block is only present in message with following types
+    if (Class == CoherenceClass:PUTX ||
+        Class == CoherenceClass:PUTX_COPY ||
+	Class == CoherenceClass:DATA ||
```

Comment:
Is there a guarantee that this data isn't stale?

Reply:
I don't have a thorough understanding of this MESI Three-Level protocol, so I'm not sure.

I added checks for `CoherenceClass:DATA` and `CoherenceClass:DATA_EXCLUSIVE` because they're similarly checked in MESI_Two_Level-msg.sm's [ResponseMsg::functionalRead()](https://github.com/gem5/gem5/blob/97a05304527f0d1112ff7aee20e22a9de11760f9/src/mem/ruby/protocol/MESI_Two_Level-msg.sm#L101), and I assumed the situation here would be analogous (although MESI_Three_Level doesn't have a separate class for requests or responses, if I'm not mistaken).

The only difference in what message types are considered from MESI Two-Level to MESI Three-Level is `CoherenceClass:PUTX_COPY`. I added that because I assumed it's similar to `CoherenceClass:PUTX`.

The message type that appears to cause the bug in #1044 is DATA_EXCLUSIVE in particular. I could revise the patch to only additionally check for that message type, rather than the other two (PUTX_COPY and DATA).

Reply:
Even if the response message's type of `DATA` implies that this data is the most up to date version in the system, I wouldn't immediately assume that it's true for the coherence message (L1->L2 message). For instance, there could be another cache that is in the process of upgrading to "M" state and already completed its store. I don't know if this implementation of MESI allows the store to complete before receiving all of the invalidation acks. It *could* depending on the consistency model, I believe.

Anyway, most of this is theoretical. I suppose if this fixes an issue and results in the correct execution it's *probably* OK. Looking at the Ruby `functionalRead` code (in `RubySystem.cc`), it seems that we need an idea of "maybe stale" in the network as well as the caches, but that doesn't exist right now.

I feel more comfortable with `DATA_EXCLUSIVE` returning the functional read than just `DATA` without a lot more digging into the details of the protocol. `PUTX` I also feel OK about. From my guess, I would say that `PUTX_COPY` is explicitly stale as the L0 cache still has the data in `M` state (according to the comment) so it could have been updated. `DATA` is *probably* ok since there is a separate `STALE_DATA` message, but that also implies that the L1 can give data to the L0 that could have a more up to date copy somewhere else, so, it's unclear if that's safe or not.